imputation/ts_impute.go

```diff
@@ -52,51 +68,134 @@ func diff(keys []string, dateFormat string) (int, int, int) {
 		}
 	}
 
-	return year, month, day
+	return parseFormat, year, month, day
+}
+
+func getSortedKeys(ts TimeSeries) []string {
+	keys := make([]string, 0)
+	for k, _ := range ts {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
+
+func getValues(ts TimeSeries) []float64 {
+	values := make([]float64, 0)
+	for _, v := range ts {
+		values = append(values, v)
+	}
+	return values
 }
 
-// FillMean returns TimeSeries with missing datapoints with value = mean(value of existing points).
+// FillNA returns TimeSeries with missing datapoints with assigned values according to the selected method.
+// Method can be "mean", "median", and "zero".
 // TODO(eftekhari-mhs): Handle error cases.
-func FillMean(ts TimeSeries) (TimeSeries, error) {
+func FillNA(ts TimeSeries, method string) (TimeSeries, error) {
 	if len(ts) < 3 {
 		log.Printf("not enough data to impute")
 		return ts, nil
 	}
-	keys := make([]string, 0)
-	var mean float64 = 0
-	for k, v := range ts {
-		keys = append(keys, k)
-		mean += v
+	keys := getSortedKeys(ts)
+	values := getValues(ts)
+
+	parseFormat, yStep, mStep, dStep := diff(keys)
+	if parseFormat == "" {
+		return ts, errors.New("date format is not ISO 8601")
 	}
-	mean = mean / float64(len(keys))
 
-	var parseFormat string
-	switch len(keys[0]) {
-	case 4:
-		parseFormat = yearfmt
-	case 7:
-		parseFormat = monthfmt
-	case 10:
-		parseFormat = dayfmt
+	log.Printf("Step is equal to : %v years, %v months, %v days \n", yStep, mStep, dStep)
+
+	// Computing the filling value.
+	var fillV float64
+	switch method {
+	case "mean":
+		for _, v := range values {
+			fillV += v
+		}
+		fillV = fillV / float64(len(keys))
+	case "zero":
+		fillV = 0
+	case "median":
+		sort.Float64s(values)
+		if len(values)%2 == 0 {
+			// In case of even length: median = average of two middle values.
+			fillV = (values[len(values)/2] + values[len(values)/2-1]) / 2
+			log.Printf("Median is %v: for even case", fillV)
+		} else {
+			fillV = values[len(values)/2]
+			log.Printf("Median is %v: for odd case", fillV)
+		}
+	default:
+		return ts, errors.New("unknown method")
 	}
 
-	if parseFormat == "" {
-		return ts, errors.New("date format is not ISO 8601")
+	// TODO(eftekhari-mhs): Handle errors.
+	startDate, _ := time.Parse(parseFormat, keys[0])
+	endDate, _ := time.Parse(parseFormat, keys[len(keys)-1])
+
+	for d := startDate; d.After(endDate) == false; d = d.AddDate(yStep, mStep, dStep) {
+		if _, ok := ts[fmt.Sprint(d.Format(parseFormat))]; !ok {
+			ts[fmt.Sprint(d.Format(parseFormat))] = fillV
+		}
+	}
+	return ts, nil
+}
+
+// Interpolate function's input is a TimeSeries with possible missing datapoints.
+// The function returns a Timeseries with no missing dates. Interpolation method is used to fill the values.
+// Methods can vary by degree: d = {1} is implemented.
+func Interpolate(ts TimeSeries, degree int) (TimeSeries, error) {
```

Comment:
It's not clear to me that FillNA and Interpolate should be separate functions. Could you convince me?

Reply:
FillNA has two inputs: time series, method = "mean", "median" but not a degree (integer)
for Interpolate we have one more input besides the method name, for now method name is assumed to be "spline" but some might consider adding "polynomial" or some other methods.
Maybe you're thinking of merging these two methods together and somehow managing one interface with two implementation? one that accept degree (int) and the other not.

Reply:
Or degree can be an optional argument. This could be an optimization for later, you can ask Jeffrey what he thinks.